src/rust/include/tracing.h

```diff
@@ -0,0 +1,252 @@
+typedef struct TracingSpanGuard TracingSpanGuard;
+
+/// Creates a span for a callsite.
+///
+/// The span must be freed when it goes out of scope.
+TracingSpanHandle* tracing_span_create(const TracingCallsite* callsite);
+
+/// Clones the given span.
+///
+/// Both spans need to be separately freed when they go out of scope.
+TracingSpanHandle* tracing_span_clone(const TracingSpanHandle* span);
+
+/// Frees a span.
+void tracing_span_free(TracingSpanHandle* span);
+
+/// Enters the given span, returning a guard.
+///
+/// `tracing_span_exit` must be called to drop this guard before the span
+/// goes out of scope.
+TracingSpanGuard* tracing_span_enter(const TracingSpanHandle* span);
+
+/// Exits a span by dropping the given guard.
+void tracing_span_exit(TracingSpanGuard* guard);
+
+/// Logs a message for a callsite.
+///
+/// You should usually call the `TracingLog` macro (or one of the helper
+/// macros such as `TracingInfo`) instead of calling this directly.
+void tracing_log(
+    const TracingCallsite* callsite,
+    const char** field_values,
+    size_t fields_len);
+
+#ifdef __cplusplus
+}
+#endif
+
+//
+// Helper macros
+//
+
+#define T_DOUBLEESCAPE(a) #a
+#define T_ESCAPEQUOTE(a) T_DOUBLEESCAPE(a)
+
+// Computes the length of the given array. This is COUNT_OF from Chromium.
+#define T_ARRLEN(x) ((sizeof(x) / sizeof(0 [x])) / ((size_t)(!(sizeof(x) % sizeof(0 [x])))))
+
+// Constructs a tracing callsite.
+//
+// name, target, and fields MUST be static constants, and the output of this
+// macro MUST be stored as a static void*.
+#define T_CALLSITE(name, target, level, fields, is_span) \
+    tracing_callsite(                                    \
+        name,                                            \
+        target,                                          \
+        level,                                           \
+        __FILE__,                                        \
+        __LINE__,                                        \
+        fields,                                          \
+        T_ARRLEN(fields),                                \
+        is_span)
+
+//
+// Spans
+//
+
+#ifdef __cplusplus
+namespace tracing
+{
+class Span;
+
+/// A guard representing a span which has been entered and is currently
+/// executing.
+///
+/// When the guard is dropped, the span will be exited.
+///
+/// This is returned by the `Span::Enter` function.
+class Entered
+{
+private:
+    friend class Span;
+    std::unique_ptr<TracingSpanGuard, decltype(&tracing_span_exit)> inner;
+
+    Entered(const TracingSpanHandle* span) : inner(tracing_span_enter(span), tracing_span_exit) {}
+};
+
+/// A handle representing a span, with the capability to enter the span if it
+/// exists.
+///
+/// If the span was rejected by the current `Subscriber`'s filter, entering the
+/// span will silently do nothing. Thus, the handle can be used in the same
+/// manner regardless of whether or not the trace is currently being collected.
```

Comment:
Yes; the filter reloading process re-evaluates the interest of all callsites in the new filter. This is incidentally why callsites must be static: they are registered in the `tracing` dispatcher to enable this kind of runtime updating, and thus need to outlive it.

Reply:
@daira the filter is initially evaluated the first time the span's callsite is hit, and then cached at the callsite so that it doesn't need to be reevaluated. 

However, the `tracing_reload` function (which is called by `setlogfilter`) [calls `ReloadHandle::reload`](https://github.com/zcash/zcash/pull/4597/files#diff-303f5e11822273f0475d52e3ee5d5c0cR148), which itself [calls a function](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing-subscriber/src/reload.rs#L148-L149) that [invalidates the cache](https://github.com/tokio-rs/tracing/blob/0da8e3f37636845ece6bdceeecb8a0946cc77076/tracing-subscriber/src/reload.rs#L168). This also applies to events (log lines).

So, any _new_ instances of a span created after the filter is reloaded will be enabled, if the filtering for that callsite changed. However, the filtering is performed _when the span is created_, so if an individual instance of a span is created when that callsite is disabled, it will remain disabled even if the callsite is enabled (the span's data will have never been recorded, for performance reasons).

Hope that's useful!

Reply:
I see. But that case would only occur if you were already in the scope of the span when you called `tracing_reload`, is that right? (If so, I think this is fine.)

Reply:
To be precise, if the span _exists_ when the filter is reloaded --- it may not be actively _entered_, since spans can be entered and exited repeatedly over their lifetime.

But, in practice, it looks like all spans in zcashd currently are only ever entered once. In general, entering a span multiple times tends to occur in asynchronous systems where tasks can yield and be resumed.